From 750fd5e8de7b2be8bd4a4b52df6fc08d2e3fb4cf Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fabian@ritter-vogt.de>
Date: Tue, 16 Apr 2024 09:49:03 +0200
Subject: [PATCH] Fall back to breeze_cursors if neither configured nor default
 can be loaded

Try harder to get some cursor theme loaded, otherwise the cursor is
invisible, which also makes it hard to fix the configuration.

Also add some warnings instead of failing silently.

(cherry picked from commit 01d224fa22de13357691f72291061323f2c1485a)
---
 src/cursor.cpp        |  5 +++++
 src/cursor.h          |  4 ++++
 src/pointer_input.cpp | 10 ++++++++++
 3 files changed, 19 insertions(+)

diff --git a/src/cursor.cpp b/src/cursor.cpp
index ab5ae731e59..62ffa6c3e5d 100644
--- a/src/cursor.cpp
+++ b/src/cursor.cpp
@@ -629,6 +629,11 @@ int Cursor::defaultThemeSize()
     return 24;
 }
 
+QString Cursor::fallbackThemeName()
+{
+    return QStringLiteral("breeze_cursors");
+}
+
 QByteArray CursorShape::name() const
 {
     switch (m_shape) {
diff --git a/src/cursor.h b/src/cursor.h
index e632b90cc21..8b53ffb81f5 100644
--- a/src/cursor.h
+++ b/src/cursor.h
@@ -148,6 +148,10 @@ public:
      * Returns the default Xcursor theme size.
      */
     static int defaultThemeSize();
+    /**
+     * Returns the fallback Xcursor theme name.
+     */
+    static QString fallbackThemeName();
 
     /**
      * Returns the current cursor position. This method does an update of the mouse position if
diff --git a/src/pointer_input.cpp b/src/pointer_input.cpp
index 1f31903f096..772c9da8de7 100644
--- a/src/pointer_input.cpp
+++ b/src/pointer_input.cpp
@@ -1053,7 +1053,17 @@ void WaylandCursorImage::updateCursorTheme()
 
     m_cursorTheme = KXcursorTheme(pointerCursor->themeName(), pointerCursor->themeSize(), targetDevicePixelRatio);
     if (m_cursorTheme.isEmpty()) {
+        qCWarning(KWIN_CORE) << "Failed to load cursor theme" << pointerCursor->themeName();
         m_cursorTheme = KXcursorTheme(Cursor::defaultThemeName(), Cursor::defaultThemeSize(), targetDevicePixelRatio);
+
+        if (m_cursorTheme.isEmpty()) {
+            qCWarning(KWIN_CORE) << "Failed to load cursor theme" << Cursor::defaultThemeName();
+            m_cursorTheme = KXcursorTheme(Cursor::fallbackThemeName(), Cursor::defaultThemeSize(), targetDevicePixelRatio);
+        }
+    }
+
+    if (m_cursorTheme.isEmpty()) {
+        qCWarning(KWIN_CORE) << "Unable to load any cursor theme";
     }
 
     Q_EMIT themeChanged();
-- 
GitLab

